### DISCUSSION TITLE
When I use Batch Optimization, will it worse than the single thread one?

### DISCUSSION THREAD

**Discussion by fengshansi** [optuna/optuna#5469] [Q&A] [answered]
The single thread one has all trial history, but the batch optimization may do not have historys of thread-1。So will it worse?
I have the same question when n_jobs>1.
I found it's log is like this:
```
[I 2024-05-30 16:19:55,427] Trial 2 finished with value: 0.7268 and parameters: {'learning_rate': 0.017185255497105728, 'depth': 1, 'subsample': 0.4593243679722948, 'colsample_bylevel': 0.9232767334934529, 'min_data_in_leaf': 29}. Best is trial 2 with value: 0.7268.
[I 2024-05-30 16:20:06,508] Trial 4 finished with value: 0.7271 and parameters: {'learning_rate': 0.0061532552090614884, 'depth': 1, 'subsample': 0.2808574143738261, 'colsample_bylevel': 0.9807914895210029, 'min_data_in_leaf': 34}. Best is trial 4 with value: 0.7271.
[I 2024-05-30 16:20:10,057] Trial 25 finished with value: 0.447 and parameters: {'learning_rate': 0.006661218611553592, 'depth': 1, 'subsample': 0.2335432895481015, 'colsample_bylevel': 0.05347637936416183, 'min_data_in_leaf': 55}. Best is trial 4 with value: 0.7271.
[I 2024-05-30 16:20:19,497] Trial 21 finished with value: 0.7268 and parameters: {'learning_rate': 0.002121250310731836, 'depth': 1, 'subsample': 0.43639176129494356, 'colsample_bylevel': 0.9892303028588831, 'min_data_in_leaf': 58}. Best is trial 4 with value: 0.7271.
[I 2024-05-30 16:20:22,485] Trial 9 finished with value: 0.7312 and parameters: {'learning_rate': 0.0225772619218575, 'depth': 2, 'subsample': 0.8229489424068829, 'colsample_bylevel': 0.2602244314929232, 'min_data_in_leaf': 32}. Best is trial 9 with value: 0.7312.
```
What dose the trial number mean? I mean I think it should start with 1 and increase in order.

**Comment by nzw0301** [comment]
Could you share the mininmal script with us?

**Reply by fengshansi** [reply]
> Could you share the mininmal script with us?

```
study = optuna.create_study(
    direction="maximize",
    sampler=TPESampler(
        seed=2024,
        multivariate=True,
        group=True,
        warn_independent_sampling=True,
        consider_magic_clip=True,
        constraints_func=None,
        categorical_distance_func=None,
        consider_endpoints=True,
        n_startup_trials=20,
    ),
)
study.optimize(
    objective,
    n_trials=100,
    show_progress_bar=True,
    n_jobs=-1,
    gc_after_trial=True,
)
```
I only used the multi-threading to accelerate. I found that TPESamlper based on trial result history. So I has a question, when I use single thread. The 50th has the result of 1-49. When I use multithread, it may only has the result of 1-48 or what, some trial before it may not complete. So I have the question, is the multithread faster and a little worse?

**Reply by nzw0301** [reply, marked as answer]
Thanks for clarifying that you are talking about `n_jobs`, not https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/009_ask_and_tell.html#batch-optimization.

> The 50th has the result of 1-49. When I use multithread, it may only has the result of 1-48 or what, some trial before it may not complete.

It depends on the trial statuses. In the 50th trial's parameter suggestions, it considers previously finished trials; at least `RUNNNING` trials are ignored.

> the multithread faster and a little worse?

It depends on the sampler's config. It happens when the sampler suggests similar params among threads. https://github.com/optuna/optuna/issues/3669#issue-1270579122 describes a little bit with nsgaii and TPEsampler's config for such a setting.
